Web Inspector: should not include implementation details of the injected script in stack traces

dcrousso opened the pull request:
#### 37f814d56d6446ea011af18ee65b6644c1127b4e
<pre>
Web Inspector: should not include implementation details of the injected script in stack traces
<a href="https://bugs.webkit.org/show_bug.cgi?id=242954">https://bugs.webkit.org/show_bug.cgi?id=242954</a>

Reviewed by Yusuke Suzuki.

* Source/JavaScriptCore/runtime/ImplementationVisibility.h:
Add a new `PrivateRecursive` that tells JSC to also propagate it to any nested functions.

* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_generator.py:
(BuiltinsGenerator.generate_embedded_code_data_for_function):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_model.py:
(BuiltinFunction.__init__):
(BuiltinFunction.fromString):
Allow builtins to define a custom `@visibility=...`. If not defined, `@linkTimeConstant` is
`Private` and everything else is `Public`.

* Source/JavaScriptCore/inspector/InjectedScriptSource.js:
Annotate `createInspectorInjectedScript` with `@visibility=PrivateRecursive`.

* Source/JavaScriptCore/parser/Nodes.h:
* Source/JavaScriptCore/parser/Nodes.cpp:
(JSC::FunctionMetadataNode::FunctionMetadataNode):
(JSC::FunctionMetadataNode::operator== const):
(JSC::FunctionMetadataNode::dump const):
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
(JSC::UnlinkedFunctionExecutable::create):
* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::createExecutable):
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.cpp:
(JSC::generateUnlinkedFunctionCodeBlock):
(JSC::UnlinkedFunctionExecutable::UnlinkedFunctionExecutable):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::makeFunction):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitNewClassFieldInitializerFunction):
Have the `FunctionMetadataNode` hold the `ImplementationVisibility` instead of directly giving it to
the `UnlinkedFunctionExecutable`, as the `FunctionMetadataNode` has more widespread usage.

* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createFunctionMetadata):
* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::Scope):
(JSC::Scope::implementationVisibility const): Added.
(JSC::Parser::pushScope):
(JSC::Parser::resetImplementationVisibilityIfNeeded): Added.
(JSC::Parser::implementationVisibility): Added.
(JSC::parse):
(JSC::parseFunctionForFunctionConstructor):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::Parser):
(JSC::Parser&lt;LexerType&gt;::isArrowFunctionParameters):
(JSC::Parser&lt;LexerType&gt;::parseGeneratorFunctionSourceElements):
(JSC::Parser&lt;LexerType&gt;::parseAsyncFunctionSourceElements):
(JSC::Parser&lt;LexerType&gt;::parseAsyncGeneratorFunctionSourceElements):
(JSC::Parser&lt;LexerType&gt;::parseFunctionBody):
(JSC::Parser&lt;LexerType&gt;::parseFunctionInfo):
* Source/JavaScriptCore/parser/SyntaxChecker.h:
(JSC::SyntaxChecker::createFunctionMetadata):
Pass along the `ImplementationVisibility` to the creation of the `FunctionMetadataNode` when parsing.
Make sure the `ImplementationVisibility` is properly propagated when new `Scope` are created.

* Source/JavaScriptCore/debugger/DebuggerParseData.cpp:
(JSC::gatherDebuggerParseData):
* Source/JavaScriptCore/runtime/CodeCache.cpp:
(JSC::generateUnlinkedCodeBlockImpl):
(JSC::CodeCache::getUnlinkedGlobalFunctionExecutable):
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::checkSyntaxInternal):
(JSC::checkModuleSyntax):
* Source/JavaScriptCore/API/JSScriptRef.cpp:
(parseScript):
* Source/JavaScriptCore/API/glib/JSCContext.cpp:
(jsc_context_check_syntax):
* Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
(JSC::moduleLoaderParseModule):
Non-builtins should be `ImplementationVisibility::Public`.

* Source/JavaScriptCore/parser/ParserModes.h:
Decrease the width of `LexicalScopeFeatures` to make room for the wider `ImplementationVisibility`.

* Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
(JSC::isAllowedReceiverFunctionForCallerAndArguments):
(JSC::RetrieveCallerFunctionFunctor::operator() const):
Ignore anything that isn&apos;t `ImplementationVisibility::Public`.

* LayoutTests/inspector/animation/lifecycle-web-animation-expected.txt:
* LayoutTests/inspector/canvas/create-context-2d-expected.txt:
* LayoutTests/inspector/canvas/create-context-bitmaprenderer-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl2-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-full-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-memoryLimit-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl-full-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl-memoryLimit-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl-snapshots-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl2-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl2-full-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl2-memoryLimit-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl2-snapshots-expected.txt:
* LayoutTests/inspector/model/remote-object/error-expected.txt:
* LayoutTests/inspector/model/stack-trace-expected.txt:
* LayoutTests/inspector/timeline/line-column-expected.txt:
* LayoutTests/platform/gtk/inspector/timeline/line-column-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-frameCount-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-full-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-memoryLimit-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-snapshots-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-frameCount-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-full-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-memoryLimit-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-snapshots-expected.txt:

Canonical link: <a href="https://commits.webkit.org/252914@main">https://commits.webkit.org/252914@main</a>
</pre>

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -220,7 +220,7 @@ static ALWAYS_INLINE bool isPrivateFieldName(UniquedStringImpl* uid)
 
 template <typename LexerType>
-Expected<typename Parser<LexerType>::ParseInnerResult, String> Parser<LexerType>::parseInner(const Identifier& calleeName, ParsingContext parsingContext, std::optional<int> functionConstructorParametersEndPosition, const FixedVector<JSTextPosition>* classFieldLocations, const PrivateNameEnvironment* parentScopePrivateNames)
+Expected<typename Parser<LexerType>::ParseInnerResult, String> Parser<LexerType>::parseInner(const Identifier& calleeName, ParsingContext parsingContext, ImplementationVisibility implementationVisibility, std::optional<int> functionConstructorParametersEndPosition, const FixedVector<JSTextPosition>* classFieldLocations, const PrivateNameEnvironment* parentScopePrivateNames)
```
We do not need to pass this parameter to every parse functions. We should store it to the FunctionScope when parsing a function. And when creating inner functions, let's get it from `currentFunctionScope()`. And from this outer function's visibility, we can get the visibility for inner functions. This scope chain may not include all scopes up to the top-level. But at least, it includes chains up to the most closest function, and that's enough.
See `ConstructorKind` for example.

Constellation replied:
For functions in Program / ModuleProgram / Eval etc., probably a bit different thing is required, but basically, putting it into scope would be the right way.

Constellation commented on Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.cpp:
```diff
@@ -46,6 +46,23 @@ static_assert(sizeof(UnlinkedFunctionExecutable) <= 128, "UnlinkedFunctionExecut
 
 const ClassInfo UnlinkedFunctionExecutable::s_info = { "UnlinkedFunctionExecutable"_s, nullptr, nullptr, nullptr, CREATE_METHOD_TABLE(UnlinkedFunctionExecutable) };
 
+static ImplementationVisibility determineImplementationVisibility(UnlinkedFunctionExecutable* executable)
```
Rename it to `implementationVisibilityForInnerFunction`.

Constellation commented on Source/JavaScriptCore/runtime/CodeCache.cpp:
```diff
@@ -248,7 +248,7 @@ UnlinkedFunctionExecutable* CodeCache::getUnlinkedGlobalFunctionExecutable(VM& v
     // The Function constructor only has access to global variables, so no variables will be under TDZ unless they're
```
We should pass ImplementationVisibility to this `getUnlinkedGlobalFunctionExecutable` function. And we should include it into `SourceCodeKey`.

Constellation replied:
Or will it be never a private visibility?

dcrousso replied:
AFAICT it seemed like all callers would expect it to be public, but I'm not super familiar with this code so I can't say for sure 😅

dcrousso replied:
looking at this again, i think sure the only caller is `constructFunction`, which i feel like should probably be public

Constellation commented on Source/JavaScriptCore/runtime/CodeCache.cpp:
```diff
@@ -77,7 +77,7 @@ UnlinkedCodeBlockType* generateUnlinkedCodeBlockImpl(VM& vm, const SourceCode& s
 
     std::unique_ptr<RootNode> rootNode = parse<RootNode>(
-        vm, source, Identifier(), JSParserBuiltinMode::NotBuiltin, strictMode, scriptMode, CacheTypes<UnlinkedCodeBlockType>::parseMode, SuperBinding::NotNeeded, error, nullptr, ConstructorKind::None, derivedContextType, evalContextType, nullptr, privateNameEnvironment, nullptr, isInsideOrdinaryFunction);
+        vm, source, Identifier(), ImplementationVisibility::Public, JSParserBuiltinMode::NotBuiltin, strictMode, scriptMode, CacheTypes<UnlinkedCodeBlockType>::parseMode, SuperBinding::NotNeeded, error, nullptr, ConstructorKind::None, derivedContextType, evalContextType, nullptr, privateNameEnvironment, nullptr, isInsideOrdinaryFunction);
```
Why is it OK to use Public? I think this function also needs to take ImplementationVisibility.

dcrousso replied:
I figured that since this was parsing with `JSParserBuiltinMode::NotBuiltin` then we'd want it to be public.  Is that wrong?

dcrousso replied:
also all of the (eventual) callers of this seemed like they'd want it public (e.g. `-[JSScript cacheBytecodeWithError:]`, `eval`, etc.)

dcrousso commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -2424,7 +2425,7 @@ template <class TreeBuilder> bool Parser<LexerType>::parseFunctionInfo(TreeBuild
             functionInfo.body = context.createFunctionMetadata(
                 startLocation, endLocation, startColumn, bodyEndColumn, 
-                functionKeywordStart, functionNameStart, parametersStart, 
+                functionKeywordStart, functionNameStart, parametersStart, implementationVisibility(),
```
@Constellation should I add this to `SourceProviderCacheItem`?  Or is it fine for this to use the current `Scope`?

Constellation replied:
I think this is fine since this status can be computed from the current scope. But it is nice to leave a comment about it here.

Constellation commented on Source/JavaScriptCore/parser/Parser.h:
```diff
@@ -1289,19 +1293,24 @@ class Parser {
         if (!m_scopeStack.isEmpty()) {
+            implementationVisibility = m_scopeStack.last().implementationVisibility();
+            if (m_scopeStack.last().isFunctionBoundary() && implementationVisibility != ImplementationVisibility::PrivateRecursive)
+                implementationVisibility = ImplementationVisibility::Public;
```
If we do this (scopes inside a function have Public while the function has Private), do we need to have `m_scopeStack.last().isFunctionBoundary() && ` part?

dcrousso replied:
ah hmm yeah good point

we need the `m_scopeStack.last().isFunctionBoundary()` part because the `Parser` adds its own top-level `Scope` and we need to make sure that the `ImplementationVisibility` propagates at least one level down

i want to move this code inside `setIsFunction()` but that doesn't have access to the `m_scopeStack`

lemme noodle on this a bit more 🤔

Constellation reviewed (approved):
r=me with comments.

dcrousso committed:
Web Inspector: should not include implementation details of the injected script in stack traces
https://bugs.webkit.org/show_bug.cgi?id=242954

Reviewed by Yusuke Suzuki.

* Source/JavaScriptCore/runtime/ImplementationVisibility.h:
Add a new `PrivateRecursive` that tells JSC to also propagate it to any nested functions.

* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_generator.py:
(BuiltinsGenerator.generate_embedded_code_data_for_function):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_model.py:
(BuiltinFunction.__init__):
(BuiltinFunction.fromString):
Allow builtins to define a custom `@visibility=...`. If not defined, `@linkTimeConstant` is
`Private` and everything else is `Public`.

* Source/JavaScriptCore/inspector/InjectedScriptSource.js:
Annotate `createInspectorInjectedScript` with `@visibility=PrivateRecursive`.

* Source/JavaScriptCore/parser/Nodes.h:
* Source/JavaScriptCore/parser/Nodes.cpp:
(JSC::FunctionMetadataNode::FunctionMetadataNode):
(JSC::FunctionMetadataNode::operator== const):
(JSC::FunctionMetadataNode::dump const):
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
(JSC::UnlinkedFunctionExecutable::create):
* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::createExecutable):
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.cpp:
(JSC::generateUnlinkedFunctionCodeBlock):
(JSC::UnlinkedFunctionExecutable::UnlinkedFunctionExecutable):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::makeFunction):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitNewClassFieldInitializerFunction):
Have the `FunctionMetadataNode` hold the `ImplementationVisibility` instead of directly giving it to
the `UnlinkedFunctionExecutable`, as the `FunctionMetadataNode` has more widespread usage.

* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createFunctionMetadata):
* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::Scope):
(JSC::Scope::implementationVisibility const): Added.
(JSC::Parser::pushScope):
(JSC::Parser::resetImplementationVisibilityIfNeeded): Added.
(JSC::Parser::implementationVisibility): Added.
(JSC::parse):
(JSC::parseFunctionForFunctionConstructor):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::Parser):
(JSC::Parser<LexerType>::isArrowFunctionParameters):
(JSC::Parser<LexerType>::parseGeneratorFunctionSourceElements):
(JSC::Parser<LexerType>::parseAsyncFunctionSourceElements):
(JSC::Parser<LexerType>::parseAsyncGeneratorFunctionSourceElements):
(JSC::Parser<LexerType>::parseFunctionBody):
(JSC::Parser<LexerType>::parseFunctionInfo):
* Source/JavaScriptCore/parser/SyntaxChecker.h:
(JSC::SyntaxChecker::createFunctionMetadata):
Pass along the `ImplementationVisibility` to the creation of the `FunctionMetadataNode` when parsing.
Make sure the `ImplementationVisibility` is properly propagated when new `Scope` are created.

* Source/JavaScriptCore/debugger/DebuggerParseData.cpp:
(JSC::gatherDebuggerParseData):
* Source/JavaScriptCore/runtime/CodeCache.cpp:
(JSC::generateUnlinkedCodeBlockImpl):
(JSC::CodeCache::getUnlinkedGlobalFunctionExecutable):
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::checkSyntaxInternal):
(JSC::checkModuleSyntax):
* Source/JavaScriptCore/API/JSScriptRef.cpp:
(parseScript):
* Source/JavaScriptCore/API/glib/JSCContext.cpp:
(jsc_context_check_syntax):
* Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
(JSC::moduleLoaderParseModule):
Non-builtins should be `ImplementationVisibility::Public`.

* Source/JavaScriptCore/parser/ParserModes.h:
Decrease the width of `LexicalScopeFeatures` to make room for the wider `ImplementationVisibility`.

* Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
(JSC::isAllowedReceiverFunctionForCallerAndArguments):
(JSC::RetrieveCallerFunctionFunctor::operator() const):
Ignore anything that isn't `ImplementationVisibility::Public`.

* LayoutTests/inspector/animation/lifecycle-web-animation-expected.txt:
* LayoutTests/inspector/canvas/create-context-2d-expected.txt:
* LayoutTests/inspector/canvas/create-context-bitmaprenderer-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl2-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-full-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-memoryLimit-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl-full-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl-memoryLimit-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl-snapshots-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl2-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl2-full-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl2-memoryLimit-expected.txt:
* LayoutTests/inspector/canvas/recording-webgl2-snapshots-expected.txt:
* LayoutTests/inspector/model/remote-object/error-expected.txt:
* LayoutTests/inspector/model/stack-trace-expected.txt:
* LayoutTests/inspector/timeline/line-column-expected.txt:
* LayoutTests/platform/gtk/inspector/timeline/line-column-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-frameCount-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-full-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-memoryLimit-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-snapshots-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-frameCount-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-full-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-memoryLimit-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-snapshots-expected.txt:

Canonical link: https://commits.webkit.org/252914@main
